### PR TITLE
Company model

### PR DESCRIPTION
**Pull request by lgarvey**
First pass at company model

### CODE REVIEW

File: company/constants.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from core.types import StrEnum
+
+
+class LastUpdatedSource(StrEnum):
+    file = 'File'
```
Comment (marcofucci) [inline company/constants.py]: I'd avoid overriding python built-ins

File: company/models.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    address_county = models.CharField(
+        _('State'),
+        blank=True,
+        max_length=30,
+    )
+
+    address_country = models.ForeignKey(
+        Country,
+        _('Country'),
+        null=True,
```
Comment (marcofucci) [inline company/models.py]: I'd have this `null=False` as records must have at least a country to be useful

File: company/models.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        related_name='+',
+    )
+
+    address_postcode = models.CharField(
+        _('Postal code'),
+        max_length=9,
```
Comment (marcofucci) [inline company/models.py]: it could be blank

File: company/models.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    registered_address_county = models.CharField(
+        _('State'),
+        blank=True,
+        max_length=30,
+    )
+
+    registered_address_country = models.ForeignKey(
+        Country,
+        _('Country'),
+        blank=True,
```
Comment (marcofucci) [inline company/models.py]: `This can be null`

File: company/models.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        _('National ID number'),
+        max_length=16,
+    )
+
+    national_id_code_type = models.CharField(
+        _('National ID code type'),
+        max_length=5,
+    )
+
+    uk_companies_house_number = models.CharField(
```
Comment (marcofucci) [inline company/models.py]: What's the difference between this and `national_id_number`?

File: company/models.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        _('UK company number'),
+        max_length=8,
+        blank=True,
+    )
+
+    line_of_business = models.CharField(
+        _('Line of business'),
+        max_length=41,
```
Comment (marcofucci) [inline company/models.py]: Could be blank (e.g. the API don't have this)

File: company/models.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        _('Glogal ultimate duns number'),
+        max_length=9,
+        blank=True,
+        validators=[RegexValidator(regex=r'^\d{9}$', message=_('Field should contain 9 numbers only'), code='invalid')],
+        db_index=True,
+    )
+
+    employee_number = models.PositiveIntegerField(
+        _('Employee number'),
+        null=True,
```
Comment (marcofucci) [inline company/models.py]: Don't you need a blank=True as well?

File: company/models.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    )
+
+    is_employees_number_estimated = models.BooleanField(
+        _('Is employee number estimated?'),
```
Comment (marcofucci) [inline company/models.py]: null=True, and blank=True?

File: company/models.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    )
+
+    annual_sales = models.FloatField(
+        _('Annual sales'),
+        help_text=_('Annual sales in US dollars'),
+        null=True,
+    )
+
+    is_annual_sales_estimated = models.BooleanField(
+        _('Is annual sales estimated?'),
```
Comment (marcofucci) [inline company/models.py]: Same as employee_number

File: company/models.py [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    )
+
+    duns_number = models.CharField(
+        _('Duns number'),
+        max_length=9,
+        db_index=True,
+        validators=[RegexValidator(regex=r'^\d{9}$', message=_('Field should contain 9 numbers only'), code='invalid')]
+    )
+
+    trading_names = ArrayField(
```
Comment (marcofucci) [inline company/models.py]: Are you thinking to incorporate the name into the trading_names? I'm not sure this is 100% accurate as the primary name might not be the trading name

File: company/models.py [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    )
+
+    LEGAL_STATUS_CHOICES = (
+        (0, _('Not available')),
+        (3, _('Corporation')),
+        (8, _('Joint venture')),
+        (12, _('Partnership')),
+        (13, _('Proprietorship')),
+        (120, _('Foreign; Company'))
+    )
```
Comment (marcofucci) [inline company/models.py]: These ones do not exist in this form in the API so they either won't be set or (more correctly) you need to model them to make them work for both flat file and API.

File: company/tests.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
Comment (marcofucci) [inline company/tests.py]: using pytest, right?

File: company/views.py [hunk 13]
```diff
@@ -0,0 +1,3 @@
+from django.shortcuts import render
+
+# Create your views here.
```
Comment (marcofucci) [inline company/views.py]: this can  go

File: company/models.py (hunk 10 above)
Comment (lgarvey) [reply company/models.py]: Ah OK, I'll add the primary name back in

File: company/models.py (hunk 3 above)
Comment (lgarvey) [reply company/models.py]: OK

File: company/models.py (hunk 4 above)
Comment (lgarvey) [reply company/models.py]: Done.

File: company/models.py (hunk 5 above)
Comment (lgarvey) [reply company/models.py]: I was contemplating populating on ingestion if national_id_code_type == 12. Happy to take your view on this though.  Does it make sense?

File: company/models.py (hunk 6 above)
Comment (lgarvey) [reply company/models.py]: Done.

File: company/models.py (hunk 7 above)
Comment (lgarvey) [reply company/models.py]: Done

File: company/models.py (hunk 8 above)
Comment (lgarvey) [reply company/models.py]: Done

File: company/models.py (hunk 9 above)
Comment (lgarvey) [reply company/models.py]: Done.

File: company/models.py (hunk 5 above)
Comment (lgarvey) [reply company/models.py]: I've ditched the uk_company_number field, and the national_id_* fields in favour of an m2m RegistrationNumber model.

File: company/models.py (hunk 11 above)
Comment (lgarvey) [reply company/models.py]: I've migrated these to enums with a partial list of choices. Can flesh out more when building the ingest code.

File: company/models.py [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    def __str__(self):
+        return self.registration_number
+
+
+class Company(models.Model):
+    """The main DNB company model"""
+
+    created = models.DateTimeField(auto_now_add=True)
+    last_updated = models.DateTimeField(auto_now=True)
+    last_updated_source = models.PositiveIntegerField(
```
Comment (marcofucci) [inline company/models.py]: PositiveIntegerField?

File: company/models.py [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+        related_name='+',
+    )
+
+    registered_address_postcode = models.CharField(
+        _('Postal code'),
+        blank=True,
+        max_length=9,
+    )
+
+    registration_numbers = models.ManyToManyField(RegistrationNumber, blank=True)
```
Comment (marcofucci) [inline company/models.py]: Wouldn't be better to have a foreignkey from `CompanyRegistationNumber` to `Company` or I'm missing something? I'm not sure you could have a registration number shared by more companies.

File: company/models.py [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+    trading_names = ArrayField(
+        models.CharField(max_length=255),
+        blank=True,
+        default=list,
+    )
+
+    address_line_1 = models.CharField(
+        _('Street address'),
+        blank=True,
+        max_length=64,
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply company/models.py]: I was going off the fields in the worldbase doc, but 255 seems like a safer bet given the API data may differ in structure.

File: company/models.py [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+    created = models.DateTimeField(auto_now_add=True)
+    last_updated = models.DateTimeField(auto_now=True)
+    last_updated_source = models.PositiveIntegerField(
+        choices=LastUpdatedSource.list()
+    )
+
+    duns_number = models.CharField(
+        _('Duns number'),
+        max_length=9,
+        db_index=True,
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply company/models.py]: Yep, this makes sense.

File: company/models.py (hunk 15 above)
Comment (lgarvey) [reply company/models.py]: I'm not sure why I would have made this an m2m, it's clearly an FK relationship.  I'll amend it.

File: company/models.py [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+    is_annual_sales_estimated = models.BooleanField(
+        _('Is annual sales estimated?'),
+        blank=True,
+        null=True,
+    )
+
+    legal_status = models.CharField(
+        _('Legal status'),
+        max_length=3,
+    )
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply company/models.py]: This is actually the legal status - which should have the LegalStatusChoices.  The field below it status_code relates to the entities org position - e.g "single location", which I think you said would be fine to skip for MVP.  I'll remove the status_code field.

File: company/models.py [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+
+    registration_type = models.CharField(
+        max_length=100,
+        choices=RegistrationNumberChoices.list()
+    )
+
+    registration_number = models.CharField(
+        max_length=50,
+        blank=False,
```
Comment (marcofucci) [inline company/models.py]: `blank=False` is the default and you have omitted it in other places so you can get rid of this I guess.

File: company/models.py [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+    )
+
+    def __str__(self):
+        return '{} / {}'.format(self.registration_type, self.registration_number)
```
Comment (marcofucci) [inline company/models.py]: f strings are awesome `f'{self.registration_type} / {self.registration_number}'` 😃

File: company/models.py [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+        Country,
+        _('Country'),
+        null=False,
+        related_name='+',
+    )
+
+    address_postcode = models.CharField(
+        _('Postal code'),
+        blank=True,
+        max_length=9,
```
Comment (marcofucci) [inline company/models.py]: Not sure all worldwide postcodes are strings of len <= 9

File: company/models.py [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+        _('Country'),
+        blank=True,
+        null=True,
+        related_name='+',
+    )
+
+    registered_address_postcode = models.CharField(
+        _('Postal code'),
+        blank=True,
+        max_length=9,
```
Comment (marcofucci) [inline company/models.py]: Same here

File: company/models.py [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+    )
+
+    line_of_business = models.CharField(
+        _('Line of business'),
+        max_length=41,
```
Comment (marcofucci) [inline company/models.py]: 41 seems pretty short but I'll leave it to you.

File: company/models.py [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+        _('Legal status'),
+        choices=LegalStatusChoices.list(),
+        max_length=100,
+    )
+
+    class Meta:
+        verbose_name_plural = 'Companies'
+
+    def __str__(self):
+        return self.primary_name
```
Comment (marcofucci) [inline company/models.py]: I'd include the duns number here as well

File: user/migrations/0001_squashed_0002_auto_20190417_0813.py [hunk 25]
```diff
@@ -39,4 +42,10 @@ class Migration(migrations.Migration):
                 ('objects', django.contrib.auth.models.UserManager()),
             ],
         ),
+        migrations.AlterModelManagers(
+            name='user',
+            managers=[
+                ('objects', user.models.UserManager()),
```
Comment (marcofucci) [inline user/migrations/0001_squashed_0002_auto_20190417_0813.py]: I think you don't really need this migration file, you can change the existing initial one by adding this `objects` as it doesn't affect the db.

File: company/models.py (hunk 21 above)
Comment (lgarvey) [reply company/models.py]: OK, I got these from the world base file doc. It makes sense to bump them though.

**Pull request merged**